### ISSUE TITLE
with_first_found will always find a task before vars in a role

### ISSUE THREAD

**Issue by JamesBelchamber** [ansible/ansible#14857]
##### Issue Type:
- Bug Report
##### Ansible Version:

```
ansible 2.0.1.0
```
##### Ansible Configuration:

None
##### Environment:

OS X
##### Summary:

When `with_first_found` is used to look up a file in `vars/` or `files/` and a file with the same name exists in `tasks/`, the file in `tasks/` will be returned instead.
##### Steps To Reproduce:

The galaxy role DavidWittman.filebeat exhibits this problem, as described in  DavidWittman/ansible-filebeat#1. This is the relevant part of the role layout:

```
├── tasks
│   ├── debian.yml
│   ├── main.yml
│   └── redhat.yml
└── vars
    ├── Debian.yml
    ├── RedHat.yml
    └── main.yml
```

And here is the play:

```
- include_vars: "{{ item }}"
  with_first_found:
    - "{{ ansible_os_family }}.yml"
    - default.yml
```

Please bear in mind that on OS X, the default filesystem is HFS+ which is case-insensitive.
##### Expected Results:

`vars/RedHat.yml` should be returned on a EL server.
##### Actual Results:

`tasks/redhat.yml` is returned and an error is reported (because it is not a vars file).

```
TASK [DavidWittman.filebeat : include_vars] ************************************
fatal: [x.x.x.x]: FAILED! => {"failed": true, "msg": "/usr/local/etc/ansible/roles/DavidWittman.filebeat/tasks/RedHat.yml must be stored as a dictionary/hash"}
```

**Comment by JamesBelchamber** [comment]
This is caused by the `path_dwim_relative` function. When `with_first_found` is used within a role, the role's path is passed to `path_dwim_relative`, which (correctly) identifies it as a role and appends the role's `tasks/` directory to the search list. This means that `tasks/` is essentially searched in every iteration. The code is [here](https://github.com/ansible/ansible/blob/4657be4eabc4d3882f407624d54715d2af197142/lib/ansible/parsing/dataloader.py#L246).

I don't really know why this behaviour is desirable so I don't know how I would go about patching this bug. It's possible that the 'patch' is really to let `path_dwim_relative` handle the list of places to search (instead of the list that's iterated through [here](https://github.com/ansible/ansible/blob/26cd4650684ce4e82fb9c6c507a23dd40718198d/lib/ansible/plugins/lookup/first_found.py#L188)), and document the order of places that `with_first_found` will search by default.

**Comment by bcoca** [comment]
this is because first_found looks at 'all' dirs no matter what the task is, it should be more discriminatory and use 'vars' in this case and not 'files' nor 'templates' which find the match in 'tasks' before even trying 'vars'

**Comment by eb4x** [comment]
Ran into this problem today aswell.

**Comment by bcoca** [comment]
please confirm that merge above fixes this

**Comment by eb4x** [comment]
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/ansible-2.2.0-py2.7.egg/ansible/executor/task_executor.py", line 92, in run
    item_results = self._run_loop(items)
  File "/Library/Python/2.7/site-packages/ansible-2.2.0-py2.7.egg/ansible/executor/task_executor.py", line 259, in _run_loop
    res = self._execute(variables=task_vars)
  File "/Library/Python/2.7/site-packages/ansible-2.2.0-py2.7.egg/ansible/executor/task_executor.py", line 454, in _execute
    result = self._handler.run(task_vars=variables)
  File "/Library/Python/2.7/site-packages/ansible-2.2.0-py2.7.egg/ansible/plugins/action/include_vars.py", line 41, in run
    source = self._find_needle('vars', source)
  File "/Library/Python/2.7/site-packages/ansible-2.2.0-py2.7.egg/ansible/plugins/action/__init__.py", line 843, in _find_needle
    result = self._loader.path_dwim_relative_stack(path_stack, dirname, needle)
  File "/Library/Python/2.7/site-packages/ansible-2.2.0-py2.7.egg/ansible/parsing/dataloader.py", line 284, in path_dwim_relative_stack
    if source.startswith('~') or source.startswith(os.path.sep):
AttributeError: 'NoneType' object has no attribute 'startswith'

**Comment by bcoca** [comment]
@eb4x that should only happen if you search for empty or null string (fixing this path now). 
@JamesBelchamber I'm going to assume everything is working for you and will close the ticket unless you tell me otherwise

**Comment by eb4x** [comment]
I've created a minimal test case you can pull, https://github.com/eb4x/ansible-bug.git

TASK [example : include_vars]
failed: [localhost](item=ansible-bug/roles/example/tasks/RedHat.yml) => {"failed": true, "item": "ansible-bug/roles/example/tasks/RedHat.yml", "message": "ansible-bug/roles/example/tasks/RedHat.yml must be stored as a dictionary/hash"}

Which makes it seem like nothing has really changed. include_vars is still pulling from tasks instead of vars directory first.

**Comment by bcoca** [comment]
so it worked for me cause of 'test bias', now it should work for everyone via ce282e93730243ebf5373ac604d5d1cb74bebd20.

**Comment by xenithorb** [comment]
Sorry to pry but:
1. I keep running into problems like this where I have to leave workarounds in my roles to actual ansible bugs. Am I using too recent of a version in Fedora? Is this issue present in all lesser versions to `2.1.1` as well?
2. What exactly _is_ the best practice way to workaround this? Understand that my mentality is that while _I could_ patch my local copy of `ansible`, I want my role to be compatible to other unsuspecting users. 
   - It seemed the most logical and elegant thing was to _not have files with the same name_, so I decided to go with `{{ ansible_os_family }}.yml` for vars, and the tasks are prefixed with `t_{{ ansible_os_family }}.yml` and are named such in their respective folders. And then I left a comment referencing this issue. Is that correct?
3. Even though this is now fixed, it's worth mentioning that the error message wasn't helpful until I came across this bug. I didn't catch on until I looked closely and saw that `include_vars:` was complaining about a file in `tasks/` versus `vars/`
